### PR TITLE
Open files in binary mode in `Header.fromfile`: fixes #8711

### LINKED ISSUE #8711: UnicodeDecodeError during fits.Header.fromfile()
MWE:
```
import numpy as np
from astropy.io import fits

fn = 'test.fits'
data = np.random.randn(1, 1)
hdu = fits.HDUList(fits.PrimaryHDU(data))
hdu.writeto(fn, overwrite=True)
header = fits.Header.fromfile(fn)
```

produces:

```
---------------------------------------------------------------------------
UnicodeDecodeError                        Traceback (most recent call last)
<ipython-input-2-32a656c8e40e> in <module>
      6 hdu = fits.HDUList(fits.PrimaryHDU(data))
      7 hdu.writeto(fn, overwrite=True)
----> 8 header = fits.Header.fromfile(fn)

/usr/local/lib/python3.6/dist-packages/astropy/io/fits/header.py in fromfile(cls, fileobj, sep, endcard, padding)
    445 
    446             return cls._from_blocks(block_iter, is_binary, sep, endcard,
--> 447                                     padding)[1]
    448         finally:
    449             if close_file:

/usr/local/lib/python3.6/dist-packages/astropy/io/fits/header.py in _from_blocks(cls, block_iter, is_binary, sep, endcard, padding)
    474         # Read the first header block.
    475         try:
--> 476             block = next(blocks)
    477         except StopIteration:
    478             raise EOFError()

/usr/local/lib/python3.6/dist-packages/astropy/io/fits/header.py in block_iter(nbytes)
    437             def block_iter(nbytes):
    438                 while True:
--> 439                     data = fileobj.read(nbytes)
    440 
    441                     if data:

/usr/lib/python3.6/codecs.py in decode(self, input, final)
    319         # decode input (taking the buffer into account)
    320         data = self.buffer + input
--> 321         (result, consumed) = self._buffer_decode(data, self.errors, final)
    322         # keep undecoded input until the next call
    323         self.buffer = data[consumed:]

UnicodeDecodeError: 'utf-8' codec can't decode byte 0xa4 in position 2883: invalid start byte
```

However, opening the file in binary mode seems to work:
```
with open(fn, 'rb') as fp:
    header = fits.Header.fromfile(fp)
```

### PR DESCRIPTION
**Pull request by astromancer**
About as small as a PR can get!

EDIT: Fix #8711

### CODE REVIEW

**Comment by pllim** [comment]
Unfortunately, this seemingly simple change broke Appveyor and 32-bit CircleCI. The HTML doc broke, I think, because of perhaps you branched off an older version of `master`.

**Comment by embray** [comment]
I'm poking at this since it's very closely related to #8707 .

Longer-term I still need to rework str/bytes handling for headers.  The current code still carries a lot of Python 2 legacy despite explicitly Python 2 support being removed.

**Comment by embray** [comment]
I think the suggestions I added in astromancer/astropy#1 will fix this.

**Comment by embray** [comment]
It would be good to add a regression test as well; perhaps more-or-less along the lines of the case you gave in #8711 would suffice.

File: astropy/io/fits/header.py [hunk 1]
```diff
@@ -2010,8 +2011,20 @@ def fromfile(cls, fileobj):
...
+            # handling; if a binary-mode file object is passed in, the user is
+            # then on their own w.r.t. newline handling.
+            #
+            # Otherwise assume we are reading from an actual FITS file and open
+            # in binary mode.
+            if sep:
+                fileobj = open(fileobj, 'r', encoding='latin1')
+            else:
+                fileobj = open(fileobj, 'rb')
+
```
Comment (saimn) [inline astropy/io/fits/header.py]: This block should be moved to the `Header.fromfile` method above. Here the method should always use bytes.

**Comment by saimn** [comment]
I rebased the branch, fixed the location of the new code, and added the changelog entry. If tests pass this should be good to go!

**Comment by bsipocz** [comment]
@saimn - can you approve it then?

**Comment by bsipocz** [comment]
Thanks @astromancer @embray and @saimn, a real teamwork!

**Comment by bsipocz** [comment]
Too bad, this has issues on LTS, too: https://travis-ci.org/astropy/astropy/jobs/546002762

Remilestoning.

**Comment by saimn** [comment]
Hmm, I guess that we would need to replace `path_like = (str, PathLike)` with `path_like = (basestring, PathLike)`. But we can probably skip this one since it fixes a not so common use case.

**Comment by bsipocz** [comment]
Yeah, I learnt the lesson a while ago of not going rouge on the fits backports. They are either clean or I should not do them.

**Pull request merged**